### PR TITLE
[BREAKING] fix isagg to correctly use a fast path

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2316

also is related to https://github.com/JuliaLang/Statistics.jl/issues/50 and https://github.com/JuliaLang/julia/issues/36978 but I work around it.

@pdeffebach - this PR uncovered many corner cases, a close look at what I propose to do would be welcome.

### CODE REVIEW

File: test/grouping.jl [hunk 1]
```diff
@@ -2691,4 +2691,122 @@ end
...
+                mean∘skipmissing, var∘skipmissing, std∘skipmissing,
+                maximum, minimum, maximum∘skipmissing, minimum∘skipmissing,
+                first, last, length, first∘skipmissing, last∘skipmissing),
+        col in ([ones(2,2), zeros(2,2), ones(2,2)], [ones(2,2), zeros(2,2), missing],
+                [DataFrame(ones(2,2)), DataFrame(zeros(2,2)), DataFrame(ones(2,2))],
+                [DataFrame(ones(2,2)), DataFrame(zeros(2,2)), ones(2,2)],
+                [DataFrame(ones(2,2)), DataFrame(zeros(2,2)), missing],
+                [(a=1, b=2), (a=3, b=4), (a=5, b=6)], [(a=1, b=2), (a=3, b=4), missing])
+        gdf = groupby_checked(DataFrame(g=[1, 1, 1], x=col), :g)
+        if fun isa typeof(length)
```
Comment (nalimilan) [inline test/grouping.jl:2797]:
```suggestion
        if fun === length
```

File: test/grouping.jl [hunk 2]
```diff
@@ -2691,4 +2691,122 @@ end
...
+            @test combine(gdf, :x => fun => :y) ≅ DataFrame(g=1, y=3)
+            @test combine(gdf, :x => (x -> fun(x)) => :y) ≅ DataFrame(g=1, y=3)
+        elseif (fun isa typeof(last) && ismissing(last(col))) ||
+               (fun isa Union{typeof(maximum), typeof(minimum)} && col ≅ [(a=1, b=2), (a=3, b=4), missing])
```
Comment (nalimilan) [inline test/grouping.jl:2801]:
```suggestion
        elseif (fun === last && ismissing(last(col))) ||
               (fun in (maximum, minimum) && col ≅ [(a=1, b=2), (a=3, b=4), missing])
```

File: test/grouping.jl [hunk 3]
```diff
@@ -2691,4 +2691,122 @@ end
...
+            # this case is a hard problem what to do and probably we have to leave it as is
```
Comment (nalimilan) [inline test/grouping.jl:2802]: I must admit I don't follow... :-p

File: test/grouping.jl [hunk 4]
```diff
@@ -2691,4 +2691,122 @@ end
...
+        end
+    end
+
+    for fun in (sum∘skipmissing, mean∘skipmissing),
+        col in ([1:3, 4:6, 7:9], [1:3, 4:6, missing])
+        gdf = groupby_checked(DataFrame(g=[1, 1, 1], x=col), :g)
+        @test combine(gdf, :x => fun => :y) ≅ combine(gdf, :x => (x -> fun(x)) => :y)
+    end
+
+    # workaround https://github.com/JuliaLang/julia/issues/36979
```
Comment (nalimilan) [inline test/grouping.jl:2760]: It's not a workaround, it's just checking that we do the same thing as Base, right?

File: src/groupeddataframe/splitapplycombine.jl [hunk 5]
```diff
@@ -761,16 +765,37 @@ end
 Reduce(f, condf=nothing, adjust=nothing) = Reduce(f, condf, adjust, false)
 
 check_aggregate(f::Any) = f
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/groupeddataframe/splitapplycombine.jl:767]: I agree. And I did it now, but it was much easier to handle them separately when developing changes :)

File: src/groupeddataframe/splitapplycombine.jl [hunk 6]
```diff
@@ -761,16 +765,37 @@ end
...
+validate_aggregate(::typeof(sum∘skipmissing), ::AbstractVector{<:Union{Missing, Number}}) = true
+
 check_aggregate(::typeof(prod)) = Reduce(Base.mul_prod)
+validate_aggregate(::typeof(prod), ::AbstractVector{<:Union{Missing, Number}}) = true
+
+check_aggregate(::typeof(prod∘skipmissing)) = Reduce(Base.mul_prod, !ismissing)
+validate_aggregate(::typeof(prod∘skipmissing), ::AbstractVector{<:Union{Missing, Number}}) = true
+
 check_aggregate(::typeof(maximum)) = Reduce(max)
+validate_aggregate(::typeof(maximum), ::AbstractVector{<:Union{Missing, Real}}) = true
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/groupeddataframe/splitapplycombine.jl:783]: > I think these methods work for any type.

I think we need this restriction:
```
julia> df = DataFrame(g=1, x=[[1,2,3], [1,2,3]])
2×2 DataFrame
│ Row │ g     │ x         │
│     │ Int64 │ Array…    │
├─────┼───────┼───────────┤
│ 1   │ 1     │ [1, 2, 3] │
│ 2   │ 1     │ [1, 2, 3] │

julia> gdf = groupby(df, :g)
GroupedDataFrame with 1 group based on key: g
First Group (2 rows): g = 1
│ Row │ g     │ x         │
│     │ Int64 │ Array…    │
├─────┼───────┼───────────┤
│ 1   │ 1     │ [1, 2, 3] │
│ 2   │ 1     │ [1, 2, 3] │

julia> combine(gdf, :x => maximum)
1×2 DataFrame
│ Row │ g     │ x_maximum │
│     │ Int64 │ Array…    │
├─────┼───────┼───────────┤
│ 1   │ 1     │ [1, 2, 3] │

julia> combine(gdf, :x => x -> maximum(x))
3×2 DataFrame
│ Row │ g     │ x_function │
│     │ Int64 │ Int64      │
├─────┼───────┼────────────┤
│ 1   │ 1     │ 1          │
│ 2   │ 1     │ 2          │
│ 3   │ 1     │ 3          │
```
but I will make the signature looser.

File: src/groupeddataframe/splitapplycombine.jl [hunk 7]
```diff
@@ -780,14 +805,36 @@ end
...
+
 check_aggregate(::typeof(std)) = Aggregate(std)
+validate_aggregate(::typeof(std), ::AbstractVector{<:Union{Missing, Number}}) = true
+
 check_aggregate(::typeof(std∘skipmissing)) = Aggregate(std, !ismissing)
+validate_aggregate(::typeof(std∘skipmissing), ::AbstractVector{<:Union{Missing, Number}}) = true
+
 check_aggregate(::typeof(first)) = Aggregate(first)
+validate_aggregate(::typeof(first), v::AbstractVector) = eltype(v) === Any ? false : true
+validate_aggregate(::typeof(first), ::AbstractVector{<:Union{Missing, MULTI_COLS_TYPE, AbstractVector}}) = false
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/groupeddataframe/splitapplycombine.jl:821]: We indeed can discuss this and this relates to what @pdeffebach wants that we allow destruction into multiple columns.

Just to be clear the `MULTI_COLS_TYPE` part is not really problematic in my opinion - it just makes sure we throw an error (as we disallow it now), so in the future we can process it correctly. This is what we try to do consistently everywhere, so that post 1.0 changes here will be non-breaking, as they will currently throw an error (this is your pet trick to handle SemVer AFAICT :smile:).

A more problematic case is `AbstractVector` as it was simply inconsistent between slow and fast paths (because fast path was not expanding it and slow path does pseudo-broadcasting).

So there actually two questions:
* if we want some `MultiCol` wrapper in the future or just allow multiple cols to be returned and silently accept it (I thought @pdeffebach wanted to silently accept this)
* do we require to opt-in for pseudo-broadcasting, we never required it in the past, and it was the default, I think we should keep it.

Although it is a legacy from the very distant past. If I were designing it now I would never expand anything neither in rows nor in columns, just store one result per group in `combine` and then say to users to use `flatten` to flatten it. But this is a completely different design in comparison to what we have now, so this is just a side comment.

File: src/groupeddataframe/splitapplycombine.jl [hunk 8]
```diff
@@ -864,7 +911,11 @@ function groupreduce_init(op, condf, adjust,
     if isconcretetype(Tnm) && applicable(initf, Tnm)
         tmpv = initf(Tnm)
         initv = op(tmpv, tmpv)
-        x = adjust isa Nothing ? initv : adjust(initv, 1)
+        if adjust isa Nothing
+            x = Tnm <: AbstractIrrational ? float(initv) : initv
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/groupeddataframe/splitapplycombine.jl:915]: I have not designed it :smile:, but actually it is inconsistent:
```
julia> zero(pi) + pi
3.141592653589793
```
and you get `Float64`. I have commented in https://github.com/JuliaLang/julia/issues/36978 to keep track of it.

File: test/grouping.jl [hunk 9]
```diff
@@ -2691,4 +2691,122 @@ end
     @test map(nrow, gdf) == [1, 1, 1]
 end
 
+@testset "isagg fix" begin
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/grouping.jl:2694]: fixed

File: test/grouping.jl [hunk 10]
```diff
@@ -2691,4 +2691,122 @@ end
...
+                mean∘skipmissing, var∘skipmissing, std∘skipmissing),
+        col in ([1, 2, 3], [big(1.5), big(2.5), big(3.5)], [1 + 0.5im, 2 + 0.5im, 3 + 0.5im],
+                [true, false, true], [pi, pi, pi], [1//2, 1//3, 1//4],
+                Real[1, 1.5, 1//2], Number[1, 1.5, 1//2], Any[1, 1.5, 1//2],
+                [1, 2, missing], [big(1.5), big(2.5), missing], [1 + 0.5im, 2 + 0.5im, missing],
+                [true, false, missing], [pi, pi, missing], [1//2, 1//3, missing],
+                Union{Missing,Real}[1, 1.5, missing],
+                Union{Missing,Number}[1, 1.5, missing], Any[1, 1.5, missing])
+        gdf = groupby_checked(DataFrame(g=[1, 1, 1], x=col), :g)
+        @test combine(gdf, :x => fun => :y) ≅ combine(gdf, :x => (x -> fun(x)) => :y)
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/grouping.jl:2705]: done

File: test/grouping.jl [hunk 11]
```diff
@@ -2691,4 +2691,122 @@ end
...
+        col in ([1, 2, 3], [big(1.5), big(2.5), big(3.5)], [1 + 0.5im, 2 + 0.5im, 3 + 0.5im],
+                [true, false, true], [pi, pi, pi], [1//2, 1//3, 1//4],
+                Real[1, 1.5, 1//2], Number[1, 1.5, 1//2], Any[1, 1.5, 1//2],
+                [1, 2, missing], [big(1.5), big(2.5), missing], [1 + 0.5im, 2 + 0.5im, missing],
+                [true, false, missing], [pi, pi, missing], [1//2, 1//3, missing],
+                Union{Missing,Real}[1, 1.5, missing],
+                Union{Missing,Number}[1, 1.5, missing], Any[1, 1.5, missing])
+        gdf = groupby_checked(DataFrame(g=[1, 1, 1], x=col), :g)
+        if fun isa typeof(last∘skipmissing)
+            # this is another hard corner case
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/grouping.jl:2730]: This is a special case, because it works in fast path, but fails in slow path. I have changed the comment

File: test/grouping.jl (hunk 4 above)
Comment (bkamins) [reply test/grouping.jl:2760]: I have changed the comment, but I would keep it. At some point https://github.com/JuliaLang/julia/issues/36979 will get fixed and then these tests will start failing so it is better to understand why then.

File: test/grouping.jl (hunk 3 above)
Comment (bkamins) [reply test/grouping.jl:2802]: I want to highlight that we do `collect` by value not by collection type, so in some cases, although a collection in general would not get accepted (it contains `NamedTuple`s on which we throw an error), it happens to be accepted because the transformation function returns a value that is accepted: `missing`.

Normally we would error like this:
```
julia> df = DataFrame(g=[1,1,2,2], x=[(a=1,b=2),(a=1,b=2), (a=1,b=2), missing])
4×2 DataFrame
│ Row │ g     │ x              │
│     │ Int64 │ NamedTuple…?   │
├─────┼───────┼────────────────┤
│ 1   │ 1     │ (a = 1, b = 2) │
│ 2   │ 1     │ (a = 1, b = 2) │
│ 3   │ 2     │ (a = 1, b = 2) │
│ 4   │ 2     │ missing        │

julia> gdf = groupby(df, :g);

julia> combine(x -> last(x.x), gdf)
ERROR: ArgumentError: function must return only single-column values, or only multiple-column values
```
but in this special case we do not.

**Comment by bkamins** [comment]
And adding checks you required caught another inconsistency - this time in `std` for `Rational` :)

**Comment by bkamins** [comment]
+ tests have shown another bug on Julia 1.0 in fast path. I handle it through type piracy, but I think it is OK, as the method is present in current releases of Julia + it will be removed when Julia 1.6 becomes LTS.

File: src/groupeddataframe/splitapplycombine.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply src/groupeddataframe/splitapplycombine.jl]: Damn, again that multi-column issue... So yeah, the fast path needs to be disabled when the column can contain `MULTI_COLS_TYPE` or `AbstractVector`.

File: src/groupeddataframe/splitapplycombine.jl [hunk 12]
```diff
@@ -1,3 +1,7 @@
+# this contstant defines which types of values returned by aggregation function
```
Comment (nalimilan) [inline src/groupeddataframe/splitapplycombine.jl:1]:
```suggestion
# this constant defines which types of values returned by aggregation function
```

File: src/groupeddataframe/splitapplycombine.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply src/groupeddataframe/splitapplycombine.jl]: Actually I had forgotten that we don't allow returning `MULTI_COLS_TYPE` currently. So at least that's safe.

`AbstractVector` remains problematic though. I wonder whether there are strong use cases for pseudo-broadcasting. It would be safer to make this opt-in for 1.0, otherwise we don't allow working with data frames whose cells contain vectors.

(BTW, instead for `MultiCol`, maybe we could reuse `AsTable`, but to wrap the returned value this time.)

File: test/grouping.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply test/grouping.jl]: Hmm, it works only with `GroupedDataFrame` input, not with a `DataFrame` input. I don't get why.

Also, I thought the goal of this PR was to make the slow and fast path behave the same, so shouldn't this always throw an error whatever the eltype?

File: src/groupeddataframe/splitapplycombine.jl [hunk 13]
```diff
@@ -1046,6 +1038,12 @@ groupreduce(f, op, condf::typeof(!ismissing), adjust, checkempty::Bool,
 (r::Reduce)(incol::AbstractVector, gd::GroupedDataFrame) =
     groupreduce((x, i) -> x, r.op, r.condf, r.adjust, r.checkempty, incol, gd)
 
+# this definition is missing in Julia 1.0 LTS and is required by aggregation for var
+# TODO: remove this when 1,0 is no longer LTS
```
Comment (nalimilan) [inline src/groupeddataframe/splitapplycombine.jl:1042]:
```suggestion
# TODO: remove this when we drop 1.0 support
```

**Comment by nalimilan** [comment]
Maybe it would be simpler to disable the fast path for `Irrational` each time we would need to special-case it? Just a thought.

File: src/groupeddataframe/splitapplycombine.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (bkamins) [reply src/groupeddataframe/splitapplycombine.jl]: I have thought about it when cleaning up the rules of pseudo-broadcasting recently. Here is what we have on 0.21:
```
julia> df = DataFrame(g=[1,2,3], x=[1:3, 4:6, 7:9])
3×2 DataFrame
│ Row │ g     │ x        │
│     │ Int64 │ UnitRan… │
├─────┼───────┼──────────┤
│ 1   │ 1     │ 1:3      │
│ 2   │ 2     │ 4:6      │
│ 3   │ 3     │ 7:9      │

julia> gdf = groupby(df, :g)
GroupedDataFrame with 3 groups based on key: g
First Group (1 row): g = 1
│ Row │ g     │ x        │
│     │ Int64 │ UnitRan… │
├─────┼───────┼──────────┤
│ 1   │ 1     │ 1:3      │
⋮
Last Group (1 row): g = 3
│ Row │ g     │ x        │
│     │ Int64 │ UnitRan… │
├─────┼───────┼──────────┤
│ 1   │ 3     │ 7:9      │

julia> combine(gdf, :x => first) # this is a wrong result and will be fixed by this PR to match what we have below
3×2 DataFrame
│ Row │ g     │ x_first  │
│     │ Int64 │ UnitRan… │
├─────┼───────┼──────────┤
│ 1   │ 1     │ 1:3      │
│ 2   │ 2     │ 4:6      │
│ 3   │ 3     │ 7:9      │

julia> combine(gdf, :x => x -> first(x)) # this is the intended output
9×2 DataFrame
│ Row │ g     │ x_function │
│     │ Int64 │ Int64      │
├─────┼───────┼────────────┤
│ 1   │ 1     │ 1          │
│ 2   │ 1     │ 2          │
│ 3   │ 1     │ 3          │
│ 4   │ 2     │ 4          │
│ 5   │ 2     │ 5          │
│ 6   │ 2     │ 6          │
│ 7   │ 3     │ 7          │
│ 8   │ 3     │ 8          │
│ 9   │ 3     │ 9          │

julia> combine(gdf, :x => Ref∘first) # this is a currently intended method to protect from unwrapping - just like in standard broadcasting
3×2 DataFrame
│ Row │ g     │ x_function │
│     │ Int64 │ UnitRange… │
├─────┼───────┼────────────┤
│ 1   │ 1     │ 1:3        │
│ 2   │ 2     │ 4:6        │
│ 3   │ 3     │ 7:9        │
```
So in short - we allow working with data frames that contain vectors as:
1. normally vectors of vectors will be returned and it is not a problem
2. if user unwraps the vector (like above - with `first`) then `Ref` can be used as in Base to protect the result

File: test/grouping.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (bkamins) [reply test/grouping.jl]: I did not see the benefit of `last∘skipmissing` throwing an error in cases when we can efficiently produce a correct result. The fact that `last∘skipmissing` errors in Base is only due to `O(1)` restriction in `last` docstring, e.g. `first∘skipmissing` works in base because its docstring does not require `O(1)`. I think we should change in Base that `last∘skipmissing` works.

> it works only with `GroupedDataFrame` input, not with a `DataFrame` input. I don't get why.

It should work in `combine` both for `GroupedDataFrame` and `DataFrame` and with `select` on `GroupedDataFrame` (and pseudo-broadcasting is applied in this case). It is expected to fail for `select` on `DataFrame`. Do you observe a different behaviour?

[the remark this answers is not in the document]
Comment (nalimilan) [reply test/grouping.jl]: Yeah, `last(skipmissing(x))` and `lastindex(skipmissing(x))` should probably be defined when `x` is an `AbstractArray`, since they don't require going over the whole collection.

> It should work in `combine` both for `GroupedDataFrame` and `DataFrame` and with `select` on `GroupedDataFrame` (and pseudo-broadcasting is applied in this case). It is expected to fail for `select` on `DataFrame`. Do you observe a different behaviour?

I get this
```julia
julia> df = DataFrame(x=[1])
1×1 DataFrame
│ Row │ x     │
│     │ Int64 │
├─────┼───────┤
│ 1   │ 1     │

julia> combine(df, :x => last ∘ skipmissing)
ERROR: MethodError: no method matching lastindex(::Base.SkipMissing{Array{Int64,1}})
```

[the remark this answers is not in the document]
Comment (bkamins) [reply test/grouping.jl]: Yeah - I have forgotten of this path. So you have:
```
julia> df = DataFrame(x=[1])
1×1 DataFrame
│ Row │ x     │
│     │ Int64 │
├─────┼───────┤
│ 1   │ 1     │

julia> combine(df, :x => last ∘ skipmissing)
ERROR: MethodError: no method matching lastindex(::Base.SkipMissing{Array{Int64,1}})

julia> combine(:x => last ∘ skipmissing, df)
1×1 DataFrame
│ Row │ x_last_skipmissing │
│     │ Int64              │
├─────┼────────────────────┤
│ 1   │ 1                  │
```
which is unfortunate.

I would not touch it though now as it is not easy to fix.

In the long term we should rewrite the whole engine for `select`/`transform`/`combine` to be unified. Now - unfortunately - I have designed it in stages, and when initially implementing `select` and `transform` we have not envisioned that we will have a full ecosystem that we have now, so we essentially have two processing engines - one for `combine` in `GroupedDataFrame` and the other for `select` in `DataFrame`.

This should be implemented in a way that `select`/`transform`/`combine` on `DataFrame` are always essentially calls to `GroupedDataFrame` on a group formed by no columns (single group). Additional things to remember when implementing it:
* add better support for 0 groups
* add possibility for a group to have zero rows
* add requested extensions to the syntax of `source_columns => function => target_col_name`

**Comment by bkamins** [comment]
> would be simpler to disable the fast path for `Irrational` each time we would need to special-case it?

I can do it if you prefer. It will be a bit simpler code, but actually more lines of code (as I have to add special methods for this case)

File: test/grouping.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply test/grouping.jl]: OK. No hurry, maybe just copy your comment to an issue to keep your analysis available.

**Comment by nalimilan** [comment]
> I can do it if you prefer. It will be a bit simpler code, but actually more lines of code (as I have to add special methods for this case, currently this is just 2 additional conditions)

Whatever you think is the cleanest.

File: src/groupeddataframe/splitapplycombine.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply src/groupeddataframe/splitapplycombine.jl]: Let's continue this discussion in a separate issue?

[the remark this answers is not in the document]
Comment (bkamins) [reply src/groupeddataframe/splitapplycombine.jl]: OK - can you please open the issue explaining what you would want to change? (given the three key considerations: 1) currently we unwrap vectors, 2) `Ref` protects, 3) in the future we want to add support for multiple columns passing)

File: test/grouping.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (bkamins) [reply test/grouping.jl]: All this is tracked in separate issues.

**Comment by bkamins** [comment]
Thank you!
